### ISSUE TITLE
[unbound-method] False positive with thisArg

### ISSUE THREAD

**Issue by mysticatea** [typescript-eslint/typescript-eslint#743]
**Repro**

```JSON
{
    "root": true,
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "project": "tsconfig.json"
    },
    "plugins": ["@typescript-eslint"],
    "rules": {
        "@typescript-eslint/unbound-method": "error"
    }
}
```

```TS
class A {
    visit(node: any): void {}
    foo(nodes: any[]) {
        nodes.forEach(this.visit, this)
    }
}
```

**Expected Result**

No errors.

**Actual Result**

```
~\dev\sandbox\test.ts
  4:23  error  Avoid referencing unbound methods which may cause unintentional scoping of `this`  @typescript-eslint/unbound-method

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

`Array#forEach` method has the second parameter as passed to `this` of the callback.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.0.0-alpha.2` |
| `@typescript-eslint/parser`        | `2.0.0-alpha.2` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.0.1` |
| `node`                             | `12.6.0` |
| `npm`                              | `6.10.1` |

**Comment by L2jLiga** [comment]
Hello, are there any updates or workarounds available?

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by RLaursen** [comment]
For those seeking a workaround, you may use an anonymous arrow function which calls a bound method, i.e.:

`nodes.forEach((node) => this.visit(node))`

I apologize if this is an unwanted bump.

**Comment by jerone** [comment]
[Playground](https://typescript-eslint.io/play/#ts=5.0.4&fileType=.tsx&code=MYGwhgzhAECC0G8CwAoa0BuBLCWAuAFAHYD2AJgKYBc0YRAngJQ0YlZmLTAlEQkgUAdCBIBzYuQqMA3NAC%2BqdADMSJCZQg069ANoBdRokXpopDYJUAnAKJhgACwJ57OQdlx4ANNGc5GxhRQ5IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

Error:
> Avoid referencing unbound methods which may cause unintentional scoping of `this`.
If your function does not access `this`, you can annotate it with `this: void`, or consider using an arrow function instead. 4:19 - 4:29

**Comment by Josh-Cena** [comment]
It has been a long time, and our latest decision is that this is an area we don't want to touch on (encoding which methods are able to provide a `this` context). It's not going to be future-proof and also not going to be consistent between built-in functions and library functions. There are many ways around it, including using an arrow function like `(x) => this.f(x)` and using a disable comment (if you really want to do this...)

In https://github.com/typescript-eslint/typescript-eslint/issues/8452 we will add to docs documenting this edge case as a known limitation.
